### DISCUSSION TITLE
Need help with creating a filter

### DISCUSSION THREAD

**Discussion by partingscientist** [uBlockOrigin/uAssets#17526] [Q&A] [answered]
On this e-commerce site:
`https://www.tokopedia.com/search?q=audio+technica+ath+m50x&st=product&navsource=home&srp_component_id=02.02.02.05`

A product (`[data-testid="divProductWrapper"]`) has ever been sold when it contains a label (`.prd_label-integrity`) stating how many of it has been sold in the past.

I tried to hide products which have never been sold using the following filter.

`tokopedia.com##[data-testid="divProductWrapper"]:not(.prd_label-integrity:upward([data-testid="divProductWrapper"])):upward(2)`

However, the above filter significantly slows down my browser. Would you mind suggesting a better filter for this case, if any exist? Thank you for your time.

**Comment by gorhill** [comment]
Not using `:upward()` inside `:not` would probably help -- currently you are forcing the `:not()` to be proecdural by putting a procedural operator in it. The following should accomplish the same thing with less overhead:

    tokopedia.com##[data-testid="divProductWrapper"]:not(.prd_label-integrity):upward([data-testid="divProductWrapper"]):upward(2)

But then there are a lot of matching items on my side, so this will always require some work.

**Reply by partingscientist** [reply]
Thank you for your help. Oddly, I find that the filter you've suggested does not match any of the nodes that I successfully targeted using the filter I previously used. I double checked this using the element picker.

**Reply by gorhill** [reply]
Yeah my bad, let me see if I can fix this.

**Reply by gorhill** [reply, marked as answer]
This should cause less overhead on browsers supporting native CSS`:has()`:

    ##[data-testid="divProductWrapper"]:not(:has(.prd_label-integrity)):upward(2)

Everything up to `:upward` is looked up using native CSS, and the final procedural `:upward` applies only to the 15 or so elements matching. I don't think it can be better than this, unless you accept to make the whole filter CSS native using:

    ##.css-12sieg3:not(:has(.prd_label-integrity))

**Reply by partingscientist** [reply]
Thank you for your help. I happen to use Firefox and I haven't enabled the flag for `:has()`. I'll experiment and consider these options. Thank you once again.
